### PR TITLE
JitArm64: Avoid loading compilerPC multiple times if it's already in a register.

### PR DESCRIPTION
**Pull request by krnlyng**
Makes the generated code slightly less redundant in some cases.

### CODE REVIEW

File: Source/Core/Core/PowerPC/JitArm64/Jit.cpp [hunk 1]
```diff
@@ -429,20 +464,24 @@ void JitArm64::WriteExit(Arm64Gen::ARM64Reg dest, bool LK, u32 exit_address_afte
   }
 }
 
-void JitArm64::FakeLKExit(u32 exit_address_after_return)
+void JitArm64::FakeLKExit(u32 exit_address_after_return, ARM64Reg preloaded_after_reg)
```
Comment (JosJuice) [inline Source/Core/Core/PowerPC/JitArm64/Jit.cpp:467]: I'm guessing the `preloaded_after_reg` name is based on the `after_reg` name, but it gets really confusing if you just read the function signature without reading the implementation of this function. At first I thought the name was talking about something that got preloaded after a register... Which doesn't really make much sense.

Ideally you should make it clear that this has a connection to the other parameter. For instance, you could name this parameter `exit_address_after_return_reg`. Though I do suppose that name is on the long side.

File: Source/Core/Core/PowerPC/JitArm64/Jit.cpp [hunk 2]
```diff
@@ -429,20 +464,24 @@ void JitArm64::WriteExit(Arm64Gen::ARM64Reg dest, bool LK, u32 exit_address_afte
...
 
   // We may need to fake the BLR stack on inlined CALL instructions.
   // Else we can't return to this location any more.
   gpr.Lock(ARM64Reg::W30);
-  ARM64Reg after_reg = gpr.GetReg();
+  ARM64Reg after_reg = preloaded_after_reg;
+  if (after_reg == ARM64Reg::INVALID_REG)
+    after_reg = gpr.GetReg();
   ARM64Reg code_reg = gpr.GetReg();
-  MOVI2R(after_reg, exit_address_after_return);
```
Comment (JosJuice) [inline Source/Core/Core/PowerPC/JitArm64/Jit.cpp:442]: Don't you still need to do this in the `after_reg == ARM64Reg::INVALID_REG` case?

File: Source/Core/Core/PowerPC/JitArm64/Jit.cpp [hunk 3]
```diff
@@ -429,20 +464,24 @@ void JitArm64::WriteExit(Arm64Gen::ARM64Reg dest, bool LK, u32 exit_address_afte
...
   ADR(EncodeRegTo64(code_reg), 12);
   STP(IndexType::Pre, EncodeRegTo64(code_reg), EncodeRegTo64(after_reg), ARM64Reg::SP, -16);
-  gpr.Unlock(after_reg, code_reg);
+  if (after_reg != preloaded_after_reg)
+    gpr.Unlock(after_reg, code_reg);
+  else
+    gpr.Unlock(code_reg);
```
Comment (JosJuice) [inline Source/Core/Core/PowerPC/JitArm64/Jit.cpp:484]: Simpler to express as:

```suggestion
  gpr.Unlock(code_reg);
  if (after_reg != preloaded_after_reg)
    gpr.Unlock(after_reg);
```

File: Source/Core/Core/PowerPC/JitArm64/Jit.cpp [hunk 4]
```diff
@@ -364,9 +365,26 @@ void JitArm64::WriteExit(u32 destination, bool LK, u32 exit_address_after_return
...
+      }
+      else
+      {
+        reg_to_push = preloaded_after_reg;
+      }
+    }
+    else
+    {
+      reg_to_push = ARM64Reg::X1;
+    }
```
Comment (JosJuice) [inline Source/Core/Core/PowerPC/JitArm64/Jit.cpp:385]: Simpler to express as:

```suggestion
    if (preloaded_after_reg == ARM64Reg::INVALID_REG)
    {
      MOVI2R(ARM64Reg::X1, exit_address_after_return);
      reg_to_push = ARM64Reg::X1;
    }
    else
    {
      reg_to_push = preloaded_after_reg;
    }
```

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_Branch.cpp [hunk 5]
```diff
@@ -168,7 +177,7 @@ void JitArm64::bcx(UGeckoInstruction inst)
   }
   else
   {
-    WriteExit(js.op->branchTo, inst.LK, js.compilerPC + 4);
+    WriteExit(js.op->branchTo, inst.LK, js.compilerPC + 4, WA);
```
Comment (JosJuice) [inline Source/Core/Core/PowerPC/JitArm64/JitArm64_Branch.cpp:180]: This is unsafe because `WA` is being passed as a scratch register to `gpr.Flush`.

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_Branch.cpp [hunk 6]
```diff
@@ -208,21 +217,23 @@ void JitArm64::bcctrx(UGeckoInstruction inst)
...
-    gpr.Unlock(WB);
   }
 
   ARM64Reg WA = gpr.GetReg();
 
   LDR(IndexType::Unsigned, WA, PPC_REG, PPCSTATE_OFF_SPR(SPR_CTR));
   AND(WA, WA, LogicalImm(~0x3, 32));
 
-  WriteExit(WA, inst.LK_3, js.compilerPC + 4);
+  WriteExit(WA, inst.LK_3, js.compilerPC + 4, WB);
```
Comment (JosJuice) [inline Source/Core/Core/PowerPC/JitArm64/JitArm64_Branch.cpp:233]: At first glance it looks like there's a bug here because you're using `WB` regardless of whether `inst.LK_3` is true... But upon closer inspection, `WriteExit` ignores the parameter if `inst.LK_3` isn't true. I don't really like it, but I'm not sure if adding an if statement or something really would be neater, so I'll leave it to your judgment.

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_Branch.cpp [hunk 7]
```diff
@@ -273,8 +284,10 @@ void JitArm64::bclrx(UGeckoInstruction inst)
     STR(IndexType::Unsigned, WB, PPC_REG, PPCSTATE_OFF_SPR(SPR_LR));
   }
 
-  gpr.Flush(conditional ? FlushMode::MaintainState : FlushMode::All, WB);
+  ARM64Reg tmp_reg = gpr.GetReg();
+  gpr.Flush(conditional ? FlushMode::MaintainState : FlushMode::All, tmp_reg);
   fpr.Flush(conditional ? FlushMode::MaintainState : FlushMode::All, ARM64Reg::INVALID_REG);
+  gpr.Unlock(tmp_reg);
```
Comment (JosJuice) [inline Source/Core/Core/PowerPC/JitArm64/JitArm64_Branch.cpp:290]: Why was this change needed?

File: Source/Core/Core/PowerPC/JitArm64/Jit.cpp [hunk 8]
```diff
@@ -450,20 +462,26 @@ void JitArm64::WriteExit(Arm64Gen::ARM64Reg dest, bool LK, u32 exit_address_afte
...
 
-void JitArm64::FakeLKExit(u32 exit_address_after_return)
+void JitArm64::FakeLKExit(u32 exit_address_after_return, ARM64Reg exit_address_after_return_reg)
 {
   if (!m_enable_blr_optimization)
     return;
 
   // We may need to fake the BLR stack on inlined CALL instructions.
   // Else we can't return to this location any more.
   gpr.Lock(ARM64Reg::W30);
```
Comment (JosJuice) [inline Source/Core/Core/PowerPC/JitArm64/Jit.cpp:472]: There's a potential for a subtle bug here if `exit_address_after_return_reg == ARM64Reg::W30`. The register cache can't correctly handle nested locks an unlocks.

I think the simplest solution would be to skip locking and unlocking W30 if `exit_address_after_return_reg == ARM64Reg::W30`. The clobbering of W30 doesn't happen until we're done using `exit_address_after_return_reg`, so it should be fine. (As long as the calling function doesn't need to use `exit_address_after_return_reg` after this function returns, which currently isn't the case, but still is a little fragile...)

Comment (JosJuice) [reply Source/Core/Core/PowerPC/JitArm64/Jit.cpp:472]: Or, alternatively: Instead of calling Lock in this function, make the caller call gpr.Lock(W30) and pass W30 as a parameter to this function. Then use that parameter as an argument for the BL (which is where W30 actually gets used).

File: Source/Core/Core/PowerPC/JitArm64/Jit.cpp [hunk 9]
```diff
@@ -450,23 +462,40 @@ void JitArm64::WriteExit(Arm64Gen::ARM64Reg dest, bool LK, u32 exit_address_afte
...
+void JitArm64::FakeLKExit(u32 exit_address_after_return, ARM64Reg exit_address_after_return_reg)
 {
   if (!m_enable_blr_optimization)
     return;
 
   // We may need to fake the BLR stack on inlined CALL instructions.
   // Else we can't return to this location any more.
-  gpr.Lock(ARM64Reg::W30);
-  ARM64Reg after_reg = gpr.GetReg();
+  if (exit_address_after_return_reg == ARM64Reg::W30)
```
Comment (JosJuice) [inline Source/Core/Core/PowerPC/JitArm64/Jit.cpp:472]:
```suggestion
  if (exit_address_after_return_reg != ARM64Reg::W30)
```

File: Source/Core/Core/PowerPC/JitArm64/Jit.cpp [hunk 10]
```diff
@@ -450,23 +462,40 @@ void JitArm64::WriteExit(Arm64Gen::ARM64Reg dest, bool LK, u32 exit_address_afte
...
+  gpr.Unlock(code_reg);
+  if (after_reg != exit_address_after_return_reg)
+    gpr.Unlock(after_reg);
 
   FixupBranch skip_exit = BL();
-  gpr.Unlock(ARM64Reg::W30);
+  if (exit_address_after_return_reg == ARM64Reg::W30)
+  {
+    gpr.Unlock(ARM64Reg::W30);
+  }
```
Comment (JosJuice) [inline Source/Core/Core/PowerPC/JitArm64/Jit.cpp:498]:
```suggestion
  if (exit_address_after_return_reg != ARM64Reg::W30)
    gpr.Unlock(ARM64Reg::W30);
```

File: Source/Core/Core/PowerPC/JitArm64/Jit.cpp [hunk 11]
```diff
@@ -450,23 +462,40 @@ void JitArm64::WriteExit(Arm64Gen::ARM64Reg dest, bool LK, u32 exit_address_afte
...
 
   // We may need to fake the BLR stack on inlined CALL instructions.
   // Else we can't return to this location any more.
-  gpr.Lock(ARM64Reg::W30);
-  ARM64Reg after_reg = gpr.GetReg();
+  if (exit_address_after_return_reg == ARM64Reg::W30)
+  {
+    // Do not lock W30 if it is the same as the exit address register, since
+    // it's already locked. It'll only get clobbered at the BL (below) where
+    // we do not need it's value anymore.
```
Comment (JosJuice) [inline Source/Core/Core/PowerPC/JitArm64/Jit.cpp:476]:
```suggestion
    // we do not need its value anymore.
```

File: Source/Core/Core/PowerPC/JitArm64/JitArm64_Branch.cpp [hunk 12]
```diff
@@ -125,6 +134,8 @@ void JitArm64::bcx(UGeckoInstruction inst)
   JITDISABLE(bJITBranchOff);
 
   ARM64Reg WA = gpr.GetReg();
+  ARM64Reg tmp_reg = inst.LK ? gpr.GetReg() : WA;
```
Comment (JosJuice) [inline Source/Core/Core/PowerPC/JitArm64/JitArm64_Branch.cpp:137]: Just a nitpick, but the code would look nicer if you renamed `tmp_reg` to WB and unlocked it at the same spot as we unlock WA.

File: Source/Core/Core/PowerPC/JitArm64/Jit.cpp (hunk 8 above)
Comment (krnlyng) [reply Source/Core/Core/PowerPC/JitArm64/Jit.cpp]: BL doesn't take register arguments, isn't it always using LR/W30, or am i misunderstanding something?

Comment (JosJuice) [reply Source/Core/Core/PowerPC/JitArm64/Jit.cpp]: Oh, I thought it did, with X30 just being the default. Alright then, let's go with the other suggestion, as you've already done.

Comment (JosJuice) [reply Source/Core/Core/PowerPC/JitArm64/Jit.cpp]: Yeah, now that I look at the instruction encoding, it has a whole 26 bits for the offset, so they couldn't fit a register in there. That's sensible.

**Review by JosJuice — APPROVED** [review]

**Comment by JosJuice** [comment]
Ah, seems like my merge of a different PR just now caused a merge conflict...

**Comment by krnlyng** [comment]
> Ah, seems like my merge of a different PR just now caused a merge conflict...

Yep looking into it.
